### ISSUE TITLE
API: timestamp resolution inference - default to one unit (if possible) instead of being data-dependent?

### ISSUE THREAD

**Issue by jorisvandenbossche** [pandas-dev/pandas#58989]
After https://github.com/pandas-dev/pandas/pull/55901, we now do inference of the best resolution, and so allow to create non-nanosecond data by default (instead of raising for out of bounds data). 

To be clear, it is a _very_ nice improvement to stop raising those OutOfBounds errors while the timestamp would perfectly fit in another resolution. But I do think we could maybe reconsider the exact logic of how to determine the resolution.

With the latest changes you get the following:

```python
>>> pd.to_datetime(["2024-03-22 11:43:01"]).dtype
dtype('<M8[s]')
>>> pd.to_datetime(["2024-03-22 11:43:01.002"]).dtype
dtype('<M8[ms]')
>>> pd.to_datetime(["2024-03-22 11:43:01.002003"]).dtype
dtype('<M8[us]')
```

The resulting dtype instance depends on the exact input value (not type). I do think this has some downsides:

- The result dtype becomes very data dependent (while in general we want to avoid value dependent behavior)
- You can very easily get multiple datetime dtypes in a workflow, causing more casting (to different unit) than necessary

The fact that pandas by default truncates the string repr of datetimes (i.e. we don't show the subsecond parts if they are all zero, regardless of the actual resolution), in contrast to numpy, also means that round-tripping through a text representation (eg CSV) will very often lead to a change in dtype.

As a potential alternative, we could also decide to have a fixed _default resolution_ (e.g. microseconds), and then the logic for inferring the resolution could be: try to use the default resolution, and only if that does not work (either out of bounds or too much precision, i.e. nanoseconds present), use the inferred resolution from the data. 

That still gives some values dependent behaviour, but I think this would make it _a lot_ less common to see. And using a resolution like microseconds is sufficient for by far most use cases (in terms of bounds it supports: [290301 BC, 294241 AD])

**Comment by jorisvandenbossche** [comment]
cc @pandas-dev/pandas-core

**Comment by WillAyd** [comment]
This sounds reasonable and I think could help simplify the implementation

**Comment by bashtage** [comment]
I think this would be an improvement.  Seems like a good idea that anyone working with human times (say down to second precision) with a range of the modern era would get the same basis for the timestamp.

**Comment by jorisvandenbossche** [comment]
> I think could help simplify the implementation

I don't think it will simplify things generally, because we still need the current inference logic when the default unit does not fit, but from looking a bit into it, I also don't think it should make the code much more complex.

**Comment by Pranav-Wadhwa** [comment]
take

**Comment by Pranav-Wadhwa** [comment]
Based on discussions, I will update [`to_datetime`](https://github.com/pandas-dev/pandas/blob/fe9ff741c4f0ed612250f2a36874406d0453f99b/pandas/core/tools/datetimes.py#L667) to always use nanoseconds in the given scenarios.

**Comment by Pranav-Wadhwa** [comment]
@jorisvandenbossche would updating the `to_datetime` function to accept a default `unit='ns'` be a feasible solution for this? Or are there cases where it wouldn't make sense to default to nanoseconds?

**Comment by WillAyd** [comment]
@Pranav-Wadhwa nanoseconds is what we used previously, so I don't think we want to go back to that. The OP suggests microseconds as a default resolution, although I'm not sure its as simple as changing the to_datetime signature either.

Before diving into the details I think should get some more agreement from the pandas core team. @jbrockmendel is our datetime guru so let's see if he has any thoughts first

**Comment by jbrockmendel** [comment]
I’m fine with OP suggestion as long as we are internally consistent, I.e. Timestamp constructor

**Comment by Pranav-Wadhwa** [comment]
@jbrockmendel what do you mean by timestamp constructor? If we set the default value of `unit='us'` in `to_datetime`, it resolves the example that the OP mentioned but would require changing many test cases in `tests/tools/test_to_datetime.py`.

**Comment by rhshadrach** [comment]
I would prefer going even further: don't automatically fallback if `us` is not suitable - require the user to manually override if they are dealing with dates beyond that range. That said, I'm still in favor of the OP as-is being what I see as a step in the right direction.

@Pranav-Wadhwa - I believe there are many places this would need to change in pandas to be consistent beyond just `to_datetime`.

**Comment by Pranav-Wadhwa** [comment]
> @Pranav-Wadhwa - I believe there are many places this would need to change in pandas to be consistent beyond just `to_datetime`.

I believe the scope of this work is outside my knowledge as this is my first issues with pandas. If it's helpful to future assignees, I found that the `to_datetime` function would default to the appropriate unit if no unit was passed in, but if `us` was passed in as the unit, it would create an object with dtype `ns` inside the `_to_datetime_with_unit` function.

**Comment by jbrockmendel** [comment]
If we do change the behavior, wouldn't it make more sense to keep "ns" as the default for backward compat?

**Comment by jorisvandenbossche** [comment]
> If we do change the behavior, wouldn't it make more sense to keep "ns" as the default for backward compat?

That would also be an option, yes. I think long term, microseconds is a better default (a larger fraction of use cases will fit in the default resolution, better compatibility with other tools), so personally I would like to switch to that at some point. But that doesn't necessarily have to happen now (e.g. could also when we would switch to a nullable extension dtype for datetime64).

I had actually started looking at this a year ago when I opened the issue, but never got around opening a PR for it. I revived that branch and got it to a basic working state -> https://github.com/pandas-dev/pandas/pull/62031. 
Currently that PR uses microseconds, but it should be relatively easy to change it to default to nanoseconds (with the logic of falling back to us when it does not fit in the range).

**Comment by rhshadrach** [comment]
> better compatibility with other tools

NumPy, PySpark, and Polars all default to microseconds, and PySpark doesn't support nano. This has been a bit of an annoyance for me personally at times (particularly PySpark). I'm positive on switching pandas over to micro, but I have no sense as to how big of a code change this would be.

**Comment by jorisvandenbossche** [comment]
> I'm positive on switching pandas over to micro, but I have no sense as to how big of a code change this would be.

For pandas itself, it is not that big, you can check it in https://github.com/pandas-dev/pandas/pull/62031 (that PR is not yet entirely complete, but already changes the default for most cases in `pd.to_datetime` and `pd.Timestamp`. Most additional code changes will be needed in updating tests)

**Comment by jorisvandenbossche** [comment]
@rhshadrach suggested at the dev meeting to make this change more gradually (because no longer inferring nanoseconds by default is a breaking change if you relied on the integer representation; although this already happens if you data source has type information that we preserve, eg reading a Parquet file that uses non-nanosecond resolution will already be read as such in pandas 2.x)

The idea would be to still default to nanoseconds (when converting from strings, stdlib datetime objecs, etc) as in pandas 2.x, but introduce a future flag to globally enable the future default of microseconds.

General thoughts on that?

Some practical questions here:

- How do we want to name this? Something like `pd.options.future.infer_microseconds = True`? (is that clear enough that it is about datetimelike data? it could also control the timedelta64 behaviour when that is implemented, so it can also be a plus that it is not too specific)
- What to do for cases that currently raise an Out-of-Bounds error for nanoseconds (for example, `pd.to_datetime("2300-01-01")`) Options:
  - Since this is currently raising an error, it is less of a backwards compatibility issue to already start using microseconds here by default. And this already enables the nice improvement to more easily handle out-of-bounds timestamps in pandas 3.0 by default (without having to enable a feature flag)
  - Keep raising an error as we do for 2.x, but then we can update the error message to point the user to enable the future flag to have it work,
- For `pd.Timestamp("..")`, the data-dependent resolution inference was actually already released in pandas 2.x. In the current PR I made for defaulting to microseconds, I also updated this constructor to consistently use microseconds (if possible). If we change to keep defaulting to nanoseconds and put the microseconds behind a future flag, I would also update the `pd.Timestamp("..")` to follow this (but this does mean a bit of back and forth for the behaviour of this constructor ..)
- If we do this, how do we test it? Add one build that enables the future default? I assume so, but given the amount of test changes currently in [#62031](https://github.com/pandas-dev/pandas/pull/62031), that might add a lot of if/else cases in the tests .. (although in some tests that are not about testing the default, we could probably fix the unit of the input data to not have it depend on the default inference)

**Comment by Dr-Irv** [comment]
My answers:

- I do think the change should be gradual.
- Maybe `pd.options.future.default_microseconds` might communicate it better?
- Don't raise if we don't have to
- I think any place where we are inferring a default resolution should be changed based on the flag.  So if `to_datetime()` is changed, then `Timestamp` and `Timedelta` should probably be changed as well.
- On testing, I looked at a few of the tests that you changed, and I'm not sure that the test needed to be changed.  If a test specified a `dtype` of `datetime64[s]`, that should work independent of this change of default resolution.  I think it would be best to really determine which test results are dependent on the default resolution (I would hope very few!), duplicate those tests to another test file, and turn on the option within that file.  I guess it would depend on how many tests are really dependent on this resolution.

**Comment by jorisvandenbossche** [comment]
> I think it would be best to really determine which test results are dependent on the default resolution (I would hope very few!),

In practice we have a lot of tests that depend on the default resolution, but just because they create test data (or expected result) using strings/timestamps/datetimes, and the conversion of that thus depends on the default resolution. But typically for what is being tested itself it does not matter, so in some cases we can indeed specify the dtype to not have it dependent on the default.

**Comment by rhshadrach** [comment]
> * Maybe `pd.options.future.default_microseconds` might communicate it better?

Agree on `default` over `infer`. I don't see it as inference when a user does `pd.Series([dt.timedelta(20)])`.

**Comment by jbrockmendel** [comment]
>  I don't see it as inference when a user does pd.Series([dt.timedelta(20)]).

nitpicky heads-up: resolution inference is implemented for datetime64 but not for timedelta64

**Comment by jbrockmendel** [comment]
IIUC the discussed behavior is "Default to nanoseconds, but give closest-available in cases where that would raise OutOfBoundsDatetime.  But also implement an option to change the default to microseconds"

I think @jorisvandenbossche mentioned in a recent dev call that actually implementing this is proving difficult.  In order to get 3.0 out the door, what if we drop the "option" part of this and just go back to nanoseconds in all not-out-of-bounds cases?  Then we can look at adding the option/deprecation for 3.x?

**Comment by jorisvandenbossche** [comment]
Sounds good. And to be clear, I don't think the option would be technically hard (essentially it is keeping the changes in my PR but putting them behind an `if` check for the option value). But adding an option is not critical for 3.0, we can also add that later for 3.1. 

The difficult part was mostly the "default to nanoseconds but use microseconds if it is out of bounds" logic, and I assume this is also certainly possible, it is just that our current conversion machinery is not set up for that out of the box (because if you default to nanoseconds from the start, and then have to fallback to microseconds, we need to keep track of whether the previous values just used nanoseconds because it was the default, or used it because it was required for the resolution of those values. That piece of information is currently not tracked, so that would have to be added)

**Comment by jbrockmendel** [comment]
I'm working on a branch to implement this.  The first sticking point:

```
ser = pd.Series(..., dtype=M8[us]")
result = ser - datetime(2016, 1, 1)
```

We cast the datetime object to a Timestamp, which in main has "us" unit.  Then `result` retains "us" unit.  In the new branch, we're the Timestamp has "ns" unit, so `result` has "ns" unit.  Which is less-good.  Everyone OK with that?

**Comment by jorisvandenbossche** [comment]
FWIW, the same issue happens right now on main if you would use a resolution of s or ms for `ser`, which then gets converted to us. 

Either we always cast stdlib `datetime` objects to the default resolution, and then this is the consequence of that. 
Or we do the casting taking the other operand's type into account. Right now the cast from datetime to Timestamp happens in `maybe_prepare_scalar_for_op` called from `IndexOpsMixin._arith_method`. We currently pass the shape of the array operand, but could also pass the dtype to take the unit into account.
 
This reminds me a bit about numpy's NEP50, where they consider Python scalars as "weakly typed" and the Python scalar gets converted to the numpy dtype when combined with a numpy array. We could do something conceptually similar here, although no idea how complicated this might become ..

(but this can probably be left for a separate discussion, and short term just accept the consequences of the current rule on main?)

**Comment by jbrockmendel** [comment]
> (but this can probably be left for a separate discussion, and short term just accept the consequences of the current rule on main?)

Sure.  In the branch I have now I've implemented the nano-with-fallback behavior for the Timestamp scalar). For vectors have reverted from infer to nano, raising OutOfBoundsDatetime for now until I can implement the fallback behavior.  In the process of updating tests.

**Comment by jbrockmendel** [comment]
Discussed again on yesterday's dev call, in particular the difficulty of implementing the fallback behavior discussed above.  I proposed an alternative that is effectively "keep the inference behavior from main, but for strings infer 'us' in cases that would previously infer 's' or 'ms'".  @jorisvandenbossche  pointed out that this was very close to his OP suggestion.  Besides the nice API, the hope was that it would be more straightforward to implement than other paths.  Participants agreed it was worth a shot  (someone wrote in the call notes "Whatever Brock can get working is fine").

It turned out to be easy to implement: #62801

**Comment by jorisvandenbossche** [comment]
I might have misunderstood the exact proposal during the call (or not have thought it properly through), but how is this supposed to address the concern from above about using microseconds instead of nanoseconds being a breaking change? (and also not yet entirely supported everywhere, eg the json issue)

**Comment by jbrockmendel** [comment]
> but how is this supposed to address the concern from above about using microseconds instead of nanoseconds being a breaking change?

It doesn't.  The proposal is a breaking change from 2.x.

**Comment by rhshadrach** [comment]
From https://github.com/pandas-dev/pandas/issues/58989#issuecomment-3453418444

> eg the json issue

What is the json issue?

**Comment by jorisvandenbossche** [comment]
> What is the json issue?

Very quick, but I think it is https://github.com/pandas-dev/pandas/issues/55827

For some reason I thought what we discussed during the dev meeting (although we discussed different options .. ;)) was not going to be a breaking change, because I seem to remember we said something like "then the json issue is less pressing and we only have to solve that for 4.0" (when we would change the default precision)

**Comment by jbrockmendel** [comment]
Yes, #55827 is the JSON issue joris referred to.  On the last dev call we discussed it briefly.  The upshot I recall was 1) we should push users away from datetimes as ints so that the ambiguity is avoided, 2) the problem exists in 2.x but is not hit often bc users don't get non-nano often, 3) it will occur more often if we change the default behaviors in 3.0, 4) reverting to all-nano for 3.0 would avoid exaserbating the issue, 5) (the part that joris and i may be remembering differently) it is not that big a deal and is OK to punt on

**Comment by jbrockmendel** [comment]
> For some reason I thought what we discussed during the dev meeting (although we discussed different options .. ;)) was not going to be a breaking change

We did discuss a bunch of options and it is easy to get wires crossed.  My concern is increasingly that we're getting bogged down again and never going to get a release out.

**Comment by jorisvandenbossche** [comment]
To avoid getting lost in details about what might have been discussed (or differently remembered) in the meetings, trying to make an explicit list of the different options that have been discussed:

1. As originally proposed in this issue, use microseconds as the default resolution in general (i.e. for inference, but when the data already has a resolution, such as numpy datetime64, we preserve that)
2. For concern of backwards compatibility, use nanoseconds as the default instead of microseconds
3. Like 2, but also already provide the automatic use of microseconds for cases that would give out-of-bounds errors for nanoseconds

The first option is what I started to implement in https://github.com/pandas-dev/pandas/pull/62031. 
Then the concern was raised that this would be a breaking change for people assuming to get nanoseconds when interpreting our datetime64 columns as integer (to be clear, this is also an issue for the current behaviour in main using data-dependent resolutions), and that we wanted to do this change of the default resolution more slowly: still default to nanoseconds for 3.0, and add a future option so we can then change the default in 4.0 (https://github.com/pandas-dev/pandas/issues/58989#issuecomment-3185400255).

Because I wanted to keep the benefit of already (automatically) getting rid of the out-of-bounds errors when parsing strings, I tried to update my PR to implement option 3. That is what I then mentioned as turning out to be more complex.

Therefore we said, let's just go back to nanoseconds for now just to get things going, i.e. option 2 (Brock's comment at https://github.com/pandas-dev/pandas/issues/58989#issuecomment-3412816537, which was thumbed up). 
Brock started to implement that, but then last week proposed to again go for microseconds because of the difficulty about the fallback (while I thought we decided we were OK to not have the fallback for now?).

---

I initially worked on moving to microseconds, so I think I am still fine with that. I just want to be explicit that I don't think going for just nanoseconds (option 2, not 3) in much harder. 

And given the previous decision to provide a slower migration away from the nanoseconds default, I would like to see some explicit support that we are now fine _not_ doing that (i.e. I think mostly @rhshadrach and @Dr-Irv argued for that in the meeting, but I might remember incorrectly)

Not yet moving to microseconds for 3.0 has the additional benefit of causing less of an issue with JSON roundtrip (https://github.com/pandas-dev/pandas/issues/55827), and also gives the possibility to change the default timedelta resolution at the same time as datetime (since timedelta right now is still nanoseconds)

**Comment by Dr-Irv** [comment]
> And given the previous decision to provide a slower migration away from the nanoseconds default, I would like to see some explicit support that we are now fine _not_ doing that (i.e. I think mostly [@rhshadrach](https://github.com/rhshadrach) and [@Dr-Irv](https://github.com/Dr-Irv) argued for that in the meeting, but I might remember incorrectly)

My main concern here is if someone has code that converts time objects to integers and assumes that those integers represent nanoseconds.

e.g. pd.Series([pd.Timestamp("2025-10-31 9:57")]).astype(int) will now return a different result.

So that makes me think we can't fully change the default - that we need a gradual transition.  On the other hand, if we document how to handle that case, I think we'd be OK.

**Comment by jbrockmendel** [comment]
> On the other hand, if we document how to handle that case, I think we'd be OK.

Where do you have in mind for this?  A sentence in the whatsnew section about this pointing out int implications?

**Comment by Dr-Irv** [comment]
> > On the other hand, if we document how to handle that case, I think we'd be OK.
> 
> Where do you have in mind for this? A sentence in the whatsnew section about this pointing out int implications?

Yes, but I also think that https://pandas.pydata.org/docs/user_guide/timeseries.html needs to be updated to indicate the newly proposed behavior.

**Comment by jbrockmendel** [comment]
Added a line in the whatsnew.  Looked in timeseries.html and the section about "From timestamps to epoch" is already robust to this change

```
We subtract the epoch (midnight at January 1, 1970 UTC) and then floor divide by the
"unit" (1 second).

.. ipython:: python

   (stamps - pd.Timestamp("1970-01-01")) // pd.Timedelta("1s")

```

**Comment by Dr-Irv** [comment]
> Added a line in the whatsnew. Looked in timeseries.html and the section about "From timestamps to epoch" is already robust to this change

I still think that file needs to be updated to indicate that the resolution you get from parsing strings varies based on the input.  And there are other places in that file where the proposed change in behavior indicates a possible doc change.  E.g., there is a section on "Representing out-of-bound spans" and  the overview talks about the pandas data type being defaulted to `ns`.

**Comment by jbrockmendel** [comment]
> I still think that file needs to be updated to indicate that the resolution you get from parsing strings varies based on the input.

Updated.

> Representing out-of-bound spans

That section is about Periods.  We can add information about non-nano support, but that can be done separately.
